src/ThreeWayMerge.php

```diff
@@ -1,57 +1,333 @@
 <?php
 
-namespace Relaxed\Merge\ThreeWayMerge;
+namespace Relaxed\Merge;
 
-use Exception;
+use Relaxed\Merge\ConflictException;
 
 class ThreeWayMerge
 {
-
     /**
-     * Performs automatic merge if no conflict arises else throws exception.
+     * Perform merge on associative array
      *
      * @param array $ancestor
      * @param array $local
      * @param array $remote
      *
      * @return array
-     * @throws Exception
      */
     public function performMerge(array $ancestor, array $local, array $remote)
     {
         // Returning a new Array for now. Can return the modified ancestor as well.
         $merged = [];
-        $conflict = false;
         foreach ($ancestor as $key => $value) {
-            if (is_array($value)) {
+            // Checks if the value contains an array itself.
+            if (is_array($value) && array_key_exists($key, $local) && array_key_exists($key, $remote)) {
                 $merged[$key] = $this->performMerge(
                     $value,
                     $local[$key],
                     $remote[$key]
                 );
             } else {
-                // If ancestor's value is equal to remote's value,
-                // Set the merged array's value to local value.
+                if (array_key_exists($key, $local) && array_key_exists($key, $remote)) {
+                    $merged[$key] = $this->merge(
+                        $ancestor[$key],
+                        $local[$key],
+                        $remote[$key],
+                        $key
+                    );
+                } elseif (array_key_exists($key, $local)) {
+                    if ($ancestor[$key] != $local[$key]) {
+                        throw new ConflictException("A conflict has occured");
+                    }
+                } elseif (array_key_exists($key, $remote)) {
+                    if ($ancestor[$key] != $remote[$key]) {
+                        throw new ConflictException("A conflict has occured");
```

Comment:
It would be kind of cool if we added some basic info about the conflict in the message. Maybe something like:
```php
throw new ConflictException("Merge conflict in $key");
```
And grammatically speking, the conflict doesn't "occur" (i.e. it doesn't happen as we parse it), it's technically already there. So let's use something like "Merge conflict in $key".